apps/managedidentity/managedidentity_test.go

```diff
@@ -1131,3 +1132,113 @@ func TestCreatingIMDSClient(t *testing.T) {
 		})
 	}
 }
+
+func TestRefreshInMultipleRequests(t *testing.T) {
+	firstToken := "first token"
+	secondToken := "new token"
+	refreshIn := 43200
+	expiresIn := 86400
+	resource := "https://resource/.default"
+	miType := SystemAssigned()
+	setEnvVars(t, CloudShell)
+
+	originalTime := getCurrentTime
+	defer func() {
+		getCurrentTime = originalTime
+	}()
+	before := cacheManager
+	defer func() { cacheManager = before }()
+	cacheManager = storage.New(nil)
+	// Create a mock client and append mock responses
+	mockClient := mock.Client{}
+	mockClient.AppendResponse(
+		mock.WithBody([]byte(fmt.Sprintf(`{"access_token":%q,"expires_in":%d,"refresh_in":%d,"token_type":"Bearer"}`, firstToken, expiresIn, refreshIn))),
+	)
+	// Create the client instance
+	client, err := New(miType, WithHTTPClient(&mockClient))
+	if err != nil {
+		t.Fatal(err)
+	}
+	ar, err := client.AcquireToken(context.Background(), resource)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Assert the first token is returned
+	if ar.AccessToken != firstToken {
+		t.Fatalf("wanted %q, got %q", firstToken, ar.AccessToken)
+	}
+
+	fixedTime := time.Now().Add(time.Duration(43400) * time.Second)
+	getCurrentTime = func() time.Time {
+		return fixedTime
+	}
+	var wg sync.WaitGroup
+	requestChecker := false
+
+	mockClient.AppendResponse(
+		mock.WithBody([]byte(fmt.Sprintf(`{"access_token":%q,"expires_in":%d,"refresh_in":%d,"token_type":"Bearer"}`, secondToken, expiresIn, refreshIn+43200))), mock.WithCallback(func(req *http.Request) {
+		}),
+	)
+	for i := 0; i < 10000; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			ar, err := client.AcquireToken(context.Background(), resource)
+			if err != nil {
+				t.Error(err)
+				return
+			}
+			if ar.AccessToken == secondToken && ar.Metadata.TokenSource == base.IdentityProvider {
+				if requestChecker {
+					t.Error("Error can only call this only once")
```

Comment:
a downside of allowing every goroutine to output an error from `go test` is that you could get 9,999 error messages and only the first would be interesting. Consider using a channel to limit the number of reported errors, [for example](https://github.com/Azure/azure-sdk-for-go/blob/01e85d2ec16d40200652aabd3e931a776deb26d8/sdk/azidentity/cache/new_test.go#L50-L68)